File: src/com/main/tasks/greedy/CoverLinesWithDots.java
package com.main.tasks.greedy;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Scanner;

import static java.lang.System.out;
import static java.util.Comparator.comparingInt;

public class CoverLinesWithDots {
    public final void start() throws IOException {
        final Long start = System.currentTimeMillis();
        prepare();
        final Long finish = System.currentTimeMillis();
        out.println("Done in: " + (finish - start) + " ms");
    }

    private void prepare() throws IOException {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        ArrayList<Line> lines = new ArrayList();
        for (int i = 0; i < n; n++) {
            int a1 = scanner.nextInt();
            int a2 = scanner.nextInt();
            lines.add(new Line(a1, a2));
        }
        scanner.close();
        solution(lines);
    }

    private void solution(ArrayList<Line> lines) {
        lines.sort(comparingInt(o -> o.to));
        ArrayList<Integer> points = new ArrayList();
        lines.forEach(line -> {
            if (points.isEmpty()) {
                points.add(line.to);
            }
            if (line.from > points.get(points.size() - 1)) {
                points.add(line.to);
            }
        });
        out.println(points.size());
        points.forEach(i -> out.print(i + " "));

    }

    public static void main(String[] args) throws IOException {
        new CoverLinesWithDots().start();
    }

    private static class Line {
        int from;
        int to;

        public Line(int f, int t) {
            from = f;
            to = t;
        }
    }
}

File: src/com/main/tasks/greedy/kotlin/CoverLinesWithDots.kt
package com.main.tasks.greedy.kotlin

import java.util.*

data class Line(val start: Int, val end: Int)

fun main() {
    with(readLines()) {
        sortBy { it.end }
        val points = fold(Stack<Int>()) { pointsStack, line ->
            pointsStack.apply {
                if (isEmpty() || line.start > peek()) push(line.end)
            }
        }
        println(points.size)
        println(points.joinToString(" "))
    }
}

fun readLines(): MutableList<Line> {
    return mutableListOf<Line>().apply {
        readlnOrNull()?.toInt()?.let {
            repeat(it) {
                val (a, b) = readln().split(" ").map { value -> value.toInt() }
                add(Line(start = a, end = b))
            }
        }
    }
}